Fix the types of C primitives and remove some that are unused

xavierleroy opened the pull request:
In preparation for a reboot of #11763 prompted by #12660, this PR makes sure that all functions declared `CAMLprim` have the correct types to be callable from the bytecode interpreter.

Some functions that are not callable from the bytecode interpreter are declared `CAMLexport` instead, or removed since they are no longer used.  See the commit messages for more details.

xavierleroy committed:
Don't define nor export as primitives the `caml_int64_*_native` functions

They were introduced in #2146 to support unboxed 64-bit integer
operations on 32-bit platforms, something that is no longer needed in
OCaml 5 and was removed in #11904.

xavierleroy committed:
Normalize the types of primitives

`caml_lex_engine`, `caml_new_lex_engine`, `caml_parse_engine` have strange
types with pointers to structs as parameters.  Give parameters
the type `value` and make explicit the casts from `value` to pointer to
struct of `value` in the function bodies.

xavierleroy committed:
Declare the arrays of primitives with the same `const`-ness as their definitions

dra27 commented on runtime/ints.c:
```diff
@@ -438,61 +438,52 @@ CAMLexport value caml_copy_int64(int64_t i)
-  if (dividend == ((int64_t)1 << 63) && divisor == -1) return dividend;
-  return dividend / divisor;
+  if (dividend == ((int64_t)1 << 63) && divisor == -1) return v1;
+  return caml_copy_int64(Int64_val(v1) / divisor);
```
```suggestion
  return caml_copy_int64(dividend / divisor);
```
or have I missed something?

xavierleroy replied:
I think you're right, but this is unrelated to this PR, is it?

xavierleroy replied:
Actually, it is related to this PR!  I just manually reverted an old commit, not noticing that it introduced this sharing of `dividend`.  Will fix shortly.

dra27 commented on runtime/runtime_events.c:
```diff
@@ -418,15 +418,15 @@ static void stw_create_runtime_events(
 }
 
-CAMLprim value caml_runtime_events_start(void) {
+CAMLprim value caml_runtime_events_start(value vunit) {
```
These three form part of the public API for runtime events, which is breaking one of the tests. Perhaps rename them to `caml_ml_runtime_events_start` (a la lots of the I/O primitives) and then either `#define caml_runtime_events_start` or an inline'd definition of `caml_runtime_events_start`?

xavierleroy replied:
You're right, it's a good idea to separate the C API functions `caml_runtime_events_start` and the primitive `caml_ml_runtime_events_start`.  I went one step further and made the C functions `void -> void` instead of `void -> value`, as `void -> void` is the natural C type for these functions.  Let me know what you think about that.

dra27 replied:
I was fairly sure I wouldn't need to suggest the change to change the return type to `void`, indeed 🙂

xavierleroy replied:
I'm glad you thought of changing the return type to `void`!

dra27 reviewed (commented):
Nice to see the end of the special case for the int64 primitives! One minor nit (possibly) in `ints.c` and some extra stuff may need to be done for the runtime_events primitives.

Is it possibly worth a one-line comment in the lexing and parsing changes to be explicit that the structs match the `value` representation, just because we don't do that elsewhere in the runtime (I think?)

xavierleroy committed:
runtime_events: introduce primitives `caml_ml_runtime_events_{start,pause,resume}`

These are wrappers for `caml_runtime_events_{start,pause,resume}` but with
types appropriate for calling from the bytecode interpreter.

xavierleroy committed:
Intermediate bootstrap while fixing the runtime_events primitives

xavierleroy committed:
runtime_events: `caml_runtime_events_{start,pause,resume}` are not primitives

They are part of the C API, but not callable from the bytecode interpreter.
Seize the opportunity to give them type `void f(void)` instead of
`value f(void)`.

xavierleroy committed:
runtime_events: `caml_runtime_events_user_resolve` is not a primitive

Its type prevents it from being called from the bytecode interpreter.

xavierleroy committed:
Bootstrap after removing some primitives

xavierleroy committed:
Changes for #12686

MisterDA commented:
Thanks for the PR, sorry for the slip-up. It looks all good to me too.
Is it worth backporting the type fixes to 4.14 and 5.1?

dra27 reviewed (approved):
LGTM!

dra27 commented:
The fix to `prims.h` in 92e16dd93ccaa3f4877e95ff1326a45af382b2b2 could be cherry-picked to 4.14/5.1; I'm not sure the rest necessarily needs to be?

xavierleroy commented:
Thanks for the review and the merge.  I don't think the wrong `extern` declarations for the arrays of primitives (without the `const` modifiers) can lead the C compiler to generate wrong code, at least until whole-program or link-time optimization is attempted, so I'd suggest not changing anything in 4.14 / 5.1 yet.  Later, if we get LTO to work safely, we may want to back-port the changes (which will include but not be limited to this PR) to 4.14 as part of long-term support.
